### PR TITLE
[PTD][Checkpoint] Add checkpointing support for DTensor submesh

### PR DESCRIPTION
**Pull request by wz337**
DTensor submesh support is added in https://github.com/pytorch/pytorch/pull/95458. 

This PR adds support for DTensor submesh by adding an extra check when create local save/load plan. 
If the rank is not participating in the mesh, we simply skip creating WriteItem/ReadItem for the local SavePlan/LoadPlan.

Updated the associated test as well.

cc. @wanchaol, @kumpera

### CODE REVIEW

File: test/distributed/checkpoint/test_dtensor_checkpoint.py [hunk 1]
```diff
@@ -79,9 +85,32 @@ def create_dtensor_model(
...
+        )
+        submesh_replicated_dt = distribute_tensor(
+            tensor_to_replicate_submesh, submesh, placements=[Replicate()]
+        )
+        model = MyTestModule(
+            sharded_dt,
+            replicated_dt,
+            submesh_sharded_dt,
+            submesh_replicated_dt,
+        ).cuda(dist.get_rank())
```
Comment (wanchaol) [inline test/distributed/checkpoint/test_dtensor_checkpoint.py:105]: nit: we don't need to put the model on cuda again, as we already put individual parameters to their own device in dtensor construction

File: torch/distributed/checkpoint/default_planner.py [hunk 2]
```diff
@@ -250,9 +250,16 @@ def create_default_local_load_plan(
...
     load requirements.
     """
+
     for fqn, obj in state_dict.items():
         md = metadata.state_dict_metadata[fqn]
-        requests += _create_read_items(fqn, md, obj)
+        # Since DTensor supports submesh, adding extra check to ensure _create_read_items()
+        # gets called only when the current rank is part of the mesh for the corresponding DTensor.
+        if isinstance(obj, DTensor):
+            if obj.device_mesh.get_coordinate() is not None:
```
Comment (wanchaol) [inline torch/distributed/checkpoint/default_planner.py:259]: curious if dcp actually writes dtensor's metadata to disk or no? i.e. device_mesh, placements, etc.

File: torch/distributed/checkpoint/default_planner.py [hunk 3]
```diff
@@ -282,8 +289,17 @@ def create_default_local_save_plan(
...
     requests = []
     for fqn, obj in state_dict.items():
-        if isinstance(obj, (ShardedTensor, DTensor)) or is_coordinator:
+        # Since DTensor supports submesh, adding extra check to ensure _create_write_items()
+        # gets called only when the current rank is part of the mesh for the corresponding DTensor.
+        if (
+            isinstance(obj, DTensor)
+            and obj.device_mesh.get_coordinate() is not None
+        ):
+            if obj.device_mesh.get_coordinate() is not None:
```
Comment (wanchaol) [inline torch/distributed/checkpoint/default_planner.py:298]: it looks like this is a duplicate `if` check?

**Review by wanchaol — APPROVED** [review]
Nice work!! Thanks for working on this, lgtm.

File: torch/distributed/checkpoint/default_planner.py (hunk 2 above)
Comment (wz337) [reply torch/distributed/checkpoint/default_planner.py:259]: Right now, the metadata looks like this. https://github.com/pytorch/pytorch/blob/master/torch/distributed/checkpoint/planner_helpers.py#L68 

The metadata was originally built around ShardedTensor, which we will probably have to revisit later on.

**Pull request closed without merging**